### PR TITLE
Simplify buildqueue

### PR DESCRIPTION
**Pull request by lucasmeijer**
- when enqueueing a node, also enqueue its use-dependencies.  much simpler, and gives better bee why output
- remove throttling feature
- stop tracking nodes specifically requested.  instead, count how many nodes have been queued. eveyr node that has ever been queued has to be finished for the build to finish
- simplify RuntimeNodes.  instead of only making RuntimeNodes for just the nodes we know could be used this run, we just allocate runtimenodes 1:1 for dagnodes.  this simplifies things as we no longer require a mapping from dagindex to runtime index.

### CODE REVIEW

File: src/DagData.hpp [hunk 1]
```diff
@@ -185,7 +185,11 @@ struct DagDerived
     uint32_t m_MagicNumber;
     uint32_t m_NodeCount;
 
-    FrozenArray<FrozenArray<int32_t>> m_Dependencies;
+
+    //this is an array of the node's direct tobuild dependencies + the usedependencies of its build dependencies.  It boils down to a flat
+    //list of everything that needs to have been produces for a node to start building.
```
Comment (mostynb) [inline src/DagData.hpp:190]: nitpick: s/produces/produced/

File: src/Main.cpp [hunk 2]
```diff
@@ -56,9 +55,6 @@ static const struct OptionTemplate
     {'D', "debug", OptionType::kBool, offsetof(DriverOptions, m_DebugMessages), "Enable debug messages"},
     {'S', "debug-signing", OptionType::kBool, offsetof(DriverOptions, m_DebugSigning), "Generate an extensive log of signature generation"},
     {'e', "just-print-leafinput-signature", OptionType::kString, offsetof(DriverOptions, m_JustPrintLeafInputSignature), "Print to the specified file the leaf input signature ingredients of the requested node"},
-    {'r', "throttle", OptionType::kBool, offsetof(DriverOptions, m_ThrottleOnHumanActivity), "Throttles down amount of simultaneous jobs when mouse or keyboard activity has been detected."},
-    {'\0', "throttle-time", OptionType::kInt, offsetof(DriverOptions, m_ThrottleInactivityPeriod), "Amount of inactive time after which we stop throttling. (if throttling behaviour is enabled)"},
-    {'\0', "throttle-threads-amount", OptionType::kInt, offsetof(DriverOptions, m_ThrottledThreadsAmount), "Amount of threads used in throttled mode"},
```
Comment (mostynb) [inline src/Main.cpp:61]: Just checking- do we use these flags on any build infra? Alternatively we could leave the flags but make them no-ops.

**Review by mostynb — APPROVED** [review]
I don't understand most of this code, but if the tests are green it's OK with me.

File: src/Main.cpp (hunk 2 above)
Comment (lucasmeijer) [reply src/Main.cpp:61]: we do not, this feature never really worked well.

File: src/Driver.cpp [hunk 3]
```diff
@@ -669,60 +666,28 @@ static void DriverSelectNodes(const Frozen::Dag *dag, const char **targets, int
...
     {
-        const Frozen::DagNode *dag_node = dag_nodes + node_indices[i];
+        const Frozen::DagNode *dag_node = dag_nodes + i;
         out_nodes[i].m_DagNode = dag_node;
-        out_nodes[i].m_DagNodeIndex = node_indices[i];
+        out_nodes[i].m_DagNodeIndex = i;
 #if ENABLED(CHECKED_BUILD)
         out_nodes[i].m_DebugAnnotation = dag_node->m_Annotation.Get();
 #endif
-        for (int j = 0; j < node_stack.m_Size; ++j)
```
Comment (jechter) [inline src/Driver.cpp:717]: Good riddance, wasn't happy with this code.

**Comment by jechter** [comment]
Nice simplifications!

**Pull request merged**